lexer/api.go

```diff
@@ -47,6 +47,11 @@ type Lexer interface {
 	Next() (Token, error)
 }
 
+// BatchLexer consumes and returns a batch of tokens at a time. A Lexer may choose to implement this for extra speed.
+type BatchLexer interface {
```

Comment:
Doesn't seem unreasonable, but how about we put this in a separate PR. Do you have any benchmarks against a lexer before/after implementing this?